src/PublicApiAnalyzers/Core/Analyzers/DeclarePublicApiAnalyzer.Impl.cs

```diff
@@ -832,7 +832,16 @@ public override bool VisitNamedType(INamedTypeSymbol symbol)
 
                     foreach (var typeArgument in symbol.TypeArguments)
                     {
-                        if (Instance.Visit(typeArgument))
+                        // type parameters will already have been checked on the type defining them, so we can just do a shallow check
+                        if (typeArgument.TypeKind == TypeKind.TypeParameter)
+                        {
+                            if (typeArgument.IsReferenceType &&
+                                typeArgument.NullableAnnotation() == NullableAnnotation.None)
+                            {
+                                return true;
+                            }
```

Comment:
❔ Why no `else return false`?

Reply:
We need to check the following type arguments (there may be more than one). Added a test to cover that scenario.